HIP/hip-0000-introduction-of-blocks.md

```diff
@@ -0,0 +1,199 @@
+    4. Getting logs based on block number filtering
+
+**Design Goal #1 Minimize Changes**
+
+The Block concept should fit naturally into the existing processes, mechanisms and state updates. It must keep the same responsibilities between consensus, services and mirror nodes.
+
+**Design Goal #2 Lightweight**
+
+The Block concept must not add a lot of complexity and performance overhead to the processing of transactions. It must have minimal impact on the TPS of the network.
+
+Based on the described design goals above, the outlined specification defines that block properties are to be computed and populated at different points in time and by different components of the network, based on their responsibility.
+
+## Specification
+
+**Definitions**
+
+- `block` → `Record file` containing all `Record Stream Objects` grouped in a 2-second time frame.
+- `block number` → consecutive number of the `Record file` that is being incremented by `1` for every new `Record file`. This value can be initially bootstrapped through `Mirror Nodes` and after that maintained by services nodes.
+- `block hash` → The `32 byte keccak256 running hash` of the `Record file`. That is the running hash of the last `Record Stream Object` from the previous `Record file`.
+- `block timestamp` → Instant of consensus timestamp of the first `transaction`/`Record Stream Object` in the `Record file`.
+
+### Platform
+
+Adapt `TimestampStreamFileWriter` to include `blockNumber` in the `Record File`. Introduce a new field `firstConsensusTimeInCurrentFile` to be used as a marker when to start a new Record file. Use the field `lastConsensusTimestamp` to keep track of the last-seen consensus timestamp that was processed. In this way, we can ensure that we have at least `1000ns`  difference between the last processed transaction before a new file is created. In this way if we have a `parent transaction` with at least one `child precompile transaction` they all will be included in the same `block`/`Record file`. Otherwise, we might have a corner case where a `parent transaction` is included in one block, and its `child precompile transaction` falls into the next block since it will increase the `consensus timestamp` with `1ns`. Therefore, the algorithm for checking whether to start a new file would be the following:
+
+A new `Record Stream Object` enters `addObject(T object)` in `TimestampStreamFileWriter`. It has a consensus timestamp `T`. We create a new file only if both (1) and (2) conditions are met:
+
+1. `T - lastConsensusTime > 1000ns`
+2. `T - firstConsensusTimeInCurrentFile > 2s`
+
+or
+
+if `lastConsensusTime` or `firstConsensusTimeInCurrentFile` is `null`
+
+### Services
+
+Services is to update the processing logic of transactions so that it supports logic for determining new 2-second periods, increment `block` number and keep `block` relevant data.
+The proposed solution specifies a `long` field to be used for the block number counter, incremented every 2 seconds.
```

Comment:
If we are going to not spell out 2 second blocks I think we need some rationale whu 2^63 is enough and 2^31 isn't for a block counter.

> Using a signed 32-bit int would result in the block number rolling over in 140 years for the initial 2 second length. Sub-second block lengths would exhaust that number well within the operational lifetime of typical networks.  A signed 64 bit integer provides a much longer timeframe.